src/api/Controllers/BoreholeGeometryController.cs

```diff
@@ -0,0 +1,317 @@
+    {
+        // Check if associated borehole is locked
+        if (await boreholeLockService.IsBoreholeLockedAsync(boreholeId, HttpContext.GetUserSubjectId()).ConfigureAwait(false))
+        {
+            return Problem("The borehole is locked by another user or you are missing permissions.", statusCode: (int)HttpStatusCode.BadRequest);
+        }
+
+        context.BoreholeGeometry.RemoveRange(context.BoreholeGeometry.Where(g => g.BoreholeId == boreholeId));
+        await context.SaveChangesAsync().ConfigureAwait(false);
+        return Ok();
+    }
+
+    [HttpGet("[action]")]
+    [Authorize(Policy = PolicyNames.Viewer)]
+    public IList<object> GeometryFormats()
+    {
+        return geometryFormats.Select(f => (object)new
+        {
+            f.Name,
+            f.Key,
+            f.CsvHeader,
+        }).ToList();
+    }
+
+    [HttpPost]
+    [Authorize(Policy = PolicyNames.Viewer)]
+    [RequestSizeLimit(int.MaxValue)]
+    [RequestFormLimits(MultipartBodyLengthLimit = MaxFileSize)]
+    public async Task<IActionResult> UploadBoreholeGeometry(int boreholeId, IFormFile geometryFile, [FromForm] string geometryFormat)
+    {
+        // Check if associated borehole is locked
+        if (await boreholeLockService.IsBoreholeLockedAsync(boreholeId, HttpContext.GetUserSubjectId()).ConfigureAwait(false))
+        {
+            return Problem("The borehole is locked by another user or you are missing permissions.", statusCode: (int)HttpStatusCode.BadRequest);
+        }
+
+        var format = geometryFormats.SingleOrDefault(f => f.Key == geometryFormat);
+        if (format == null)
+        {
+            return Problem("Invalid geometry format.", statusCode: (int)HttpStatusCode.BadRequest);
+        }
+
+        // convert geometry data to BoreholeGeometry
+        IList<BoreholeGeometryElement> boreholeGeometry;
+        try
+        {
+            boreholeGeometry = format.ReadCsv(geometryFile, boreholeId);
+        }
+        catch (Exception ex) when (ex is HeaderValidationException || ex is ReaderException || ex is TypeConverterException)
+        {
+            return Problem(ex.Message, statusCode: (int)HttpStatusCode.BadRequest);
+        }
+
+        // delete existing geometry data of borehole
+        context.BoreholeGeometry.RemoveRange(context.BoreholeGeometry.Where(g => g.BoreholeId == boreholeId));
+
+        // add new geometry data to database
+        await context.BoreholeGeometry.AddRangeAsync(boreholeGeometry).ConfigureAwait(false);
+
+        try
+        {
+            await context.SaveChangesAsync().ConfigureAwait(false);
+            return Ok();
+        }
+        catch (Exception ex)
+        {
+            var errorMessage = "An error occurred while saving the borehole geometry.";
+            logger?.LogError(ex, errorMessage);
+            return Problem(errorMessage);
+        }
+    }
+
+    private interface IBoreholeGeometryFormat
```

Comment:
PP/REQ: Wie kurz angesprochen finde ich die Interfaces und Klassen hier fehl am Platz, da sie nichts mit dem Controller zu tun haben bzw. man sie nicht hier suchen würde. Daher sollten alle Klassen in separate Files in einen eigenen (Unter-)Namespace verschoben werden. 

Falls es zu viele Files gibt, bin ich auch mit folgendem Kompromiss einverstanden: Alle Interfaces und Klassen in ein einziges File ausserhalb des Controllers packen und das File aussagekräftig benennen.

Reply:
Was mich hier besonders stört ist die inkonsistenz. Der UploadController müsste man demnach ebenfalls überarbeiten.

Reply:
@patrickackermann Da bin ich bei dir. Siehe https://github.com/swisstopo/swissgeol-boreholes-suite/issues/662

Reply:
@patrickackermann Du kannst diesen Schritt sonst auch in einem separaten PR nachliefern.